ssl/s3_lib.c

```diff
@@ -4723,12 +4735,27 @@ EVP_PKEY *ssl_generate_pkey_group(SSL *s, uint16_t id)
                  ERR_R_EVP_LIB);
         goto err;
     }
-    if (gtype != TLS_CURVE_CUSTOM
+# ifndef OPENSSL_NO_DH
+    if (gtype == TLS_GROUP_FFDHE) {
+        if ((pkey = EVP_PKEY_new()) == NULL
+                || !EVP_PKEY_assign(pkey, EVP_PKEY_DH, DH_new_by_nid(ginf->nid))
+                || EVP_PKEY_CTX_set_dh_nid(pctx, ginf->nid) <= 0) {
+            SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_SSL_GENERATE_PKEY_GROUP,
+                    ERR_R_EVP_LIB);
+            EVP_PKEY_free(pkey);
```

Comment:
I don't think here double free occurs for `pkey`. If I am not correct please clarify me. Because `EVP_PKEY_assign` assigns return value of `DH_new_by_nid` to `pkey`. And when `EVP_PKEY_CTX_set_dh_nid` fails freeing `pkey` is valid as per my view.

Reply:
Ah...yeah you're right. I was getting pkey and pctx mixed up. However I think there may be a potential mem leak. I think the DH object leaks if EVP_PKEY_assign fails?

Reply:
No, I debugged and verified this also. In `EVP_PKEY_assign` before assigning `DH*` a call to `EVP_PKEY_set_type` is done which assigns `ameth`. So in case if `EVP_PKEY_CTX_set_dh_nid` fails, then `EVP_PKEY_free` cleans `pkey->pkey.ptr` if `ameth` is initialized. Debugged using `gdb` and verified that `int_dh_free` is called in this place.

Reply:
I think you didn't quite understand my point. I am concerned with the case where `EVP_PKEY_assign` fails. So in that case `DH_new_by_nid()` will create a new DH object and pass it to `EVP_PKEY_assign` - but because that function fails the pointer does not get stored in `pkey->pkey.ptr` and therefore does not get freed by `EVP_PKEY_free`.

Reply:
Ya. Restructured the checks to fix this, please check.